### ISSUE TITLE
Semantic segmentation with foreground input 

### ISSUE THREAD

**Issue by Wapity** [MIC-DKFZ/nnUNet#415]
Hello again Fabian, 
I would like to ask you a more domain-related question this time, do you know how I could train a semantic segmentation to discriminate between foreground classes directly given the foreground mask? I tried penalizing the cross entropy with 0 as weight for the background but without much result. Also by masking the output. I think of concatenating it with the input. Is there a way to force the network to predict only 2 channels (and remove the channel for background) ? It seems it is taking always background into account even if removed from the dataset.json, and it is comprehensible. 
Thanks a lot !

**Comment by FabianIsensee** [comment]
Hi,
I am not sure what you ware trying to achieve? You have an image that only contains foreground? Why would nnU-Net not work in this case? It would learn to never predict background in that case

**Comment by Wapity** [comment]
Hello, 
Sorry if I was not clear, I want to do a semantic segmentation between two classes of lesions, my labels are 0:background, 1:class A, 2:class B, but I have already the agnostic mask composed of all lesions, so it could help me to use this information, and force the network on the simpler task of disctinction between the two rather than segmentation from background and between them.
Concatenating the mask seems to help, If I want to put all background to 1 so it only predicts 1 or 2 you think I can do it directly in the dataloader or I need to reprocess the data for it? as would be taken into account for the expected output of the nnunet of 3 channels here (and I would like only 2) 
Thank you

**Comment by FabianIsensee** [comment]
I am not sure what the best solution is. There is no easy way of concatenating segmentation masks with data without touching the code. What You could try is repurpose the nnUNetTrainerV2BraTSRegions trainer to make it predict background, lesion (as in both classes A and B), class A and class B

**Comment by Wapity** [comment]
Thank you, is it because the nnUNetTrainerV2BraTSRegions tries to determine the foreground before doing the multi class split? 
I think I suceeded adding the mask in `run_iteration` of the nnUnetTrainerV2 with ` data = torch.cat([data, torch.clamp(target[0], max=1.)], dim=1)` and it seems it is working, getting 0.9 of dice for class 1 and 0.2 of dice for class 2 (very unbalanced dataset), you think it is well the right place? it seems to go directly in the network without preprocessing of the mask 
I then tried to add it for the final prediction of whole patient, and added the mask in `predict.py` after 'starting prediction' in the loop of preprocessed : 
`mask_path = output_filename.replace('inference_output', 'inference_input_mask')[
            :-7] + '_0002.nii.gz'
        mask = np.expand_dims(nib.load(mask_path).get_fdata(
        ).transpose((2, 0, 1)), axis=0)
        dims = d.shape
        mask = mask[:, :dims[1], :dims[2], :dims[3]]
        d = np.concatenate((d, mask))`
So the reconstruction is well done by the sliding window approach, however the final result seems like mirrored, is mirror not only for data augmentation? there is also the bbox cropping, the final output is well of shape 180x256x256 as input 
Would you know why? I join you images 

Final thing, I saw there is an argument do_bg in GDL loss, it allows to compute the loss only on foreground if set to false? 
Thank you very much for your help, I try to see the complex whole puzzle, impressive work but hard for me to get into 
<img width="156" alt="mask_origin" src="https://user-images.githubusercontent.com/28689667/101341786-e6db4b00-3881-11eb-921d-329f50697353.PNG">
<img width="157" alt="mask_prediction" src="https://user-images.githubusercontent.com/28689667/101341790-e773e180-3881-11eb-8452-e49f7204d5d3.PNG">

**Comment by Wapity** [comment]
Actually I found for the case of rotation, it was how I loaded the nifti file compared to how we do it here

**Comment by FabianIsensee** [comment]
Yeah that is unfortunaltey still not an entirely solved problem. nnU-Net disregards all geometry information except the spacing. Your pixel arrays must be aligned, not the geometries
